### PR TITLE
Get rid of Engine Concept Cache

### PR DESCRIPTION
**Pull request by fppt**

### CODE REVIEW

**Comment by fppt** [comment]
please retest this please

**Comment by fppt** [comment]
DO NOT MERGE

**Comment by fppt** [comment]
please retest this please - ai.grakn.test.engine.GraknEngineFailoverIT.whenSubmittingTasksToOneEngineAndRandomlyKillingTheOthers_TheyComplete

File: grakn-core/src/main/java/ai/grakn/graph/admin/GraknAdmin.java [hunk 1]
```diff
@@ -125,26 +126,11 @@
...
-    void clear(ConceptCache conceptCache);
-
-    /**
      * Commits to the graph without submitting any commit logs.
-     *
+     * @return the commit log that would have been submitted if it is needed.
      * @throws GraknValidationException when the graph does not conform to the object concept
      */
-    void commitNoLogs() throws GraknValidationException;
+    Optional<String> commitNoLogs() throws GraknValidationException;
```
Comment (alexandraorth) [inline grakn-core/src/main/java/ai/grakn/graph/admin/GraknAdmin.java]: Not a to-do for now, but just to note: we should turn this into some other structured object in the future.

File: grakn-engine/src/main/java/ai/grakn/engine/postprocessing/PostProcessing.java [hunk 2]
```diff
@@ -173,37 +134,22 @@ private void waitToContinue() {
...
-        }
-
-        LOG.info("    Total Jobs: " + total);
+    /**
+     * Functional interface for the perform post processing methods
+     * @param <A> the keyspace
+     * @param <B> the index of the concept
+     * @param <C> the set of concept ids to post process
+     */
+    @FunctionalInterface
```
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/postprocessing/PostProcessing.java]: We should delete this. (I have code that already does, which I can submit as a PR once this is in central)

File: grakn-engine/src/main/java/ai/grakn/engine/postprocessing/PostProcessing.java [hunk 3]
```diff
@@ -51,22 +49,17 @@
  */
 public class PostProcessing {
```
Comment (alexandraorth) [inline grakn-engine/src/main/java/ai/grakn/engine/postprocessing/PostProcessing.java]: We dont really need this class at all now - can eventually move the whole thing into `PostProcessingTask`

File: grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java [hunk 4]
```diff
@@ -940,6 +912,10 @@ public boolean fixDuplicateCastings(String index, Set<ConceptId> castingVertexId
             //Remove Redundant Relations
             duplicateRelations.forEach(relation -> ((ConceptImpl) relation).deleteNode());
 
+            //Restore the index
+            String newIndex = mainCasting.getIndex();
```
Comment (alexandraorth) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: What is this for?

File: grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java [hunk 5]
```diff
@@ -852,6 +833,29 @@ public void commit(BiConsumer<Set<Pair<String, ConceptId>>, Set<Pair<String,Conc
...
+        commitTransactionInternal();
+
+        //TODO: Kill when analytics no longer needs this
+        GraknSparkComputer.refresh();
+
+        LOG.trace("Graph committed.");
+
+        //No post processing should ever be done for the system keyspace
+        if(!keyspace.equalsIgnoreCase(SystemKeyspace.SYSTEM_GRAPH_NAME) && submissionNeeded) {
+            return Optional.of(conceptLog.toString());
```
Comment (alexandraorth) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: If no PP should ever be done for the system keyspace, should we ever return the logs?

File: grakn-core/src/main/java/ai/grakn/graph/admin/GraknAdmin.java (hunk 1 above)
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graph/admin/GraknAdmin.java]: Agreed.

File: grakn-engine/src/main/java/ai/grakn/engine/postprocessing/PostProcessing.java (hunk 3 above)
Comment (fppt) [inline grakn-engine/src/main/java/ai/grakn/engine/postprocessing/PostProcessing.java]: Agreed again but I would say do that in another PR

File: grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java (hunk 5 above)
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: We don't return the logs if it has to do with the system graph. That's what that check is for. When the system graph is committed an empty optional is returned.

File: grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java (hunk 4 above)
Comment (fppt) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: This is because deleting the duplicate castings deleted the index. This puts the index back on. Without this we end up with corrupted indices. This was a particularly fun one to solve

File: grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java (hunk 5 above)
Comment (alexandraorth) [inline grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java]: Right, of course, I was reading the if backwards.

**Comment by alexandraorth** [comment]
I'm going to rebase this on master and re-submit as my own PR (as Filipe is on holiday).

**Comment by bolerio** [comment]
@alexandraorth by close, do you mean "accept"? :)
